### PR TITLE
tls: add certificate expiry stats for each TlsContext

### PR DESCRIPTION
**Pull request by Shikugawa**
Commit Message: tls: add certificate expiry stats for each TlsContext
Additional Description: This PR adds new stat (`listener.<envoy_listener_address>.tls_context.<stat_prefix>.days_until_cert_expiring`) to show certificate expiration days for each TLS context.
In Envoy, `days_until_first_cert_expiring` works fine to identify whether there are any certificate to notice expiration. But it is not enough if there are many certificates bound for the single Envoy instance. Therefore, we need to have another one to identify the certificate which expires soon.

Risk Level: Low
Testing: Integration
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Comment by tyxia** [comment]
Friendly ping @adisuissa on API review.  Also assign @ggreenway as code/domain owner

@Shikugawa please also merge main to resolve the conflict

File: api/envoy/extensions/transport_sockets/tls/v3/common.proto [hunk 1]
```diff
@@ -251,6 +251,10 @@ message TlsCertificate {
 
   // [#not-implemented-hide:]
   repeated config.core.v3.DataSource signed_certificate_timestamp = 5;
+
+  // Optional. The prefix to use for stats related with this certificate.
```
Comment (adisuissa) [inline api/envoy/extensions/transport_sockets/tls/v3/common.proto:255]: nit:
```suggestion
  // The prefix to use for stats related with this certificate.
```

**Review by adisuissa — COMMENTED** [review]
Thanks for adding this!
A high-level comment: In most of the places I'm aware of `stat_prefix` only adds a prefix to some stats that are omitted (i.e., the optionality here is not whether stats are omitted or not, but what will be the stat-prefix, if any).
Consider changing this to behave the same to:
1) make it more consistent across the board
2) decouple emitting a stat from its name (prefix)

**Comment by Shikugawa** [comment]
> Thanks for adding this! A high-level comment: In most of the places I'm aware of `stat_prefix` only adds a prefix to some stats that are omitted (i.e., the optionality here is not whether stats are omitted or not, but what will be the stat-prefix, if any). Consider changing this to behave the same to:
> 
> 1. make it more consistent across the board
> 2. decouple emitting a stat from its name (prefix)

@adisuissa I also wondered how to identify the stat for each TLS context. I agree with the idea that this name is not consistent with other stat-prefix usage. I think `name` could be better (But I think that `name` is too generic)

**Comment by ggreenway** [comment]
This has been worked on before in https://github.com/envoyproxy/envoy/pull/27687. Please read conversation/comments in that PR for direction on how this should look.

/wait

**Comment by Shikugawa** [comment]
@ggreenway Oh. I couldn't find the previous work related with this. And I roughly caught the discussion. I think we should fix the following points in this PR, right?

- Show the expiration date since unix epoch start instead of expiration duration

**Comment by ggreenway** [comment]
> @ggreenway Oh. I couldn't find the previous work related with this. And I roughly caught the discussion. I think we should fix the following points in this PR, right?
> 
> * Show the expiration date since unix epoch start instead of expiration duration

Yes, that's the biggest difference. Do you think that's workable?

**Comment by Shikugawa** [comment]
@ggreenway Yes. It sounds reasonable and easy. Let me work on it.

**Pull request closed without merging**